src/box/txn.c

```diff
@@ -553,17 +559,40 @@ txn_begin_in_engine(struct engine *engine, struct txn *txn)
 {
 	if (engine->flags & ENGINE_BYPASS_TX)
 		return 0;
-	if (txn->engine == NULL) {
-		txn->engine = engine;
-		return engine_begin(engine, txn);
-	} else if (txn->engine != engine) {
-		/**
-		 * Only one engine can be used in
-		 * a multi-statement transaction currently.
-		 */
-		diag_set(ClientError, ER_CROSS_ENGINE_TRANSACTION);
+	if (txn->engines[engine->id] != NULL) {
+		assert(txn_has_flag(txn, TXN_IS_STARTED_IN_ENGINE));
+		assert(txn->engines[engine->id] == engine);
+		return 0;
+	}
+	if (!txn_has_flag(txn, TXN_IS_STARTED_IN_ENGINE)) {
+		txn_set_flags(txn, TXN_IS_STARTED_IN_ENGINE);
+		assert(txn_has_flag(txn, TXN_SUPPORTS_MULTI_ENGINE));
+		if (!engine_supports_cross_engine_tx(engine))
+			txn_clear_flags(txn, TXN_SUPPORTS_MULTI_ENGINE);
+		goto begin_in_engine;
+	}
```

Comment:
`engine_begin()` may fail, in which case the transaction flags will be wrong. You should update the flags only on success.

Reply:
Fixed.